homework-tk

tatiana-kornienko14 opened the pull request.

Tatyana145 committed:
дз Корниенко

tatiana-kornienko14 committed:
Update utils.cpp

tatiana-kornienko14 committed:
Update utils.cpp

LostPointer commented on homework_01/task_01/src/utils.cpp:
```diff
@@ -3,5 +3,24 @@
 
 std::vector<std::string> SplitString(const std::string& data) {
-  return {};
+  std::vector<std::string> v;
```
плохое имя у переменной, непонятно что она означает

LostPointer commented on homework_01/task_01/src/utils.cpp:
```diff
@@ -3,5 +3,24 @@
+  std::vector<std::string> v;
+  std::string tmp = "";
+
+  for (auto c : s) {
```
тут ошибка компиляции будет, нет переменной s

LostPointer commented on homework_01/task_01/src/utils.cpp:
```diff
@@ -3,5 +3,24 @@
+  std::string tmp = "";
+
+  for (auto c : s) {
+    if (c != ' ') {
```
если посмотреть в тесты, то там не только по пробелу разделение идет, но и по символу табуляции ('\t')

LostPointer commented on homework_01/task_02/src/utils.cpp:
```diff
@@ -1,7 +1,101 @@
+#include <iostream>
```
```#include "utils.hpp"

#include <iostream>
#include <regex>
#include <stack>```

обычно так оформляют, сначала файл где объявлены функции которые определены в этом файле, потом инклуды стандартной библиотеки, потом инклуды сторонних библиотек/фреймворков, и потом файлы из проекта

LostPointer commented on homework_01/task_02/src/utils.cpp:
```diff
@@ -1,7 +1,101 @@
-
+#include <regex>
 #include <stack>
+#define vector_string std::vector<std::string>
```
```using vector_string = std::vector<std::string>;```

так же можно использовать typedef

не стоит использовать define где этого можно избежать

LostPointer commented on homework_01/task_02/src/utils.cpp:
```diff
@@ -1,7 +1,101 @@
 
 int Calculate(const std::string& data) {
-  return 0;
+    std::vector<int> v1; // хотела сделать тип double, но в тестах int, так что пока так
```
если хочешь сделай double, это только плюс

название переменной так себе (непонятно что означает)

LostPointer commented on homework_01/task_02/src/utils.cpp:
```diff
@@ -1,7 +1,101 @@
-  return 0;
+    std::vector<int> v1; // хотела сделать тип double, но в тестах int, так что пока так
+    std::string tmp;
+    bool f = 0, pr = 0, dl = 0, mns = 0;
```
совсем непонятные названия переменных, и хорошая практика: одна переменная - одна строка

LostPointer commented on homework_01/task_02/src/utils.cpp:
```diff
@@ -1,7 +1,101 @@
+    bool f = 0, pr = 0, dl = 0, mns = 0;
+    
+    for (size_t i = 0; i < data.size(); ++i) {
+        if (data[i] != "*" && data[i] != "+" && data[i] != "-" && data[i] != "/" && !(is_numb(data[i]))) {
```
!(is_numb(data[i])) лишние круглые скобки

а ещё лучше вот так оформить:
```
  if (data[i] != "*" &&
      data[i] != "+" &&
      data[i] != "-" &&
      data[i] != "/" &&
      !is_numb(data[i]))
```

ilya-kadochnikov replied:
Следующий код не скомпилируется
`if (data[i] != "*" && data[i] != "+" && data[i] != "-" && data[i] != "/" && !(is_numb(data[i])))`.
Первая причина: `data[i]` имеет тип `char`, а символы "*", "+", "-" и "/" тип `const char*`, и сравнивать их нельзя.
Вторая причина: функция `is_numb` принимает аргумент типа `const std::string &str`, а вы передаёте `char`.
Можно исправить эту строчку так:
`if (data[i] != '*' && data[i] != '+' && data[i] != '-' && data[i] != '/' && !(is_numb(std::to_string(data[i]))))`

LostPointer commented on homework_01/task_02/src/utils.cpp:
```diff
@@ -1,7 +1,101 @@
+    
+    for (size_t i = 0; i < data.size(); ++i) {
+        if (data[i] != "*" && data[i] != "+" && data[i] != "-" && data[i] != "/" && !(is_numb(data[i]))) {
+            std::cout << "ошибка!\n";
```
тут можно кинуть исключение, например [std::runtime_error](https://en.cppreference.com/w/cpp/error/runtime_error)

LostPointer commented:
в первой задаче не рассмотрен случай со скобками и табуляцией

ilya-kadochnikov commented on homework_01/task_02/src/utils.cpp:
```diff
@@ -1,7 +1,101 @@
 #include <stack>
+#define vector_string std::vector<std::string>
+
+bool is_numb(const std::string &str) {
```
В большинстве случаев не стоит сокращать слова в именах. Имя `IsNumber` (ну или `is_number`) проще понять, чем `is_numb` (к тому же numb переводится с английского как онемевший, что добавляет двусмысленности)

ilya-kadochnikov commented on homework_01/task_02/src/utils.cpp:
```diff
@@ -1,7 +1,101 @@
+        }
+
+        if (f) {
+            if (data[i] == ")") {
```
Не скомпилируется. Исправление: `data[i] == ')'`

ilya-kadochnikov commented on homework_01/task_02/src/utils.cpp:
```diff
@@ -1,7 +1,101 @@
+            continue;
+        }
+
+        if (data[i] == "(") {
```
`data[i] == '('`

ilya-kadochnikov commented on homework_01/task_02/src/utils.cpp:
```diff
@@ -1,7 +1,101 @@
+        }
+
+        if (pr) {
+            v1[v1.size()-1] *= std::stoi(data[i]);
```
Не скомпилируется. Функция `std::stoi` принимает строку, не `char`. Вариант исправления: 
```
v1[v1.size() - 1] *= data[i] - '0';
```

ilya-kadochnikov commented on homework_01/task_02/src/utils.cpp:
```diff
@@ -1,7 +1,101 @@
+        }
+
+        if (dl) {
+            v1[v1.size()-1] /= std::stoi(data[i]);
```
`v1[v1.size() - 1] /= data[i] - '0'`

ilya-kadochnikov commented on homework_01/task_02/src/utils.cpp:
```diff
@@ -1,7 +1,101 @@
+            continue;
+        }
+
+        if (data[i] == "*") {
```
`data[i] == '*'`

ilya-kadochnikov commented on homework_01/task_02/src/utils.cpp:
```diff
@@ -1,7 +1,101 @@
+            continue;
+        }
+
+        if (data[i] == "/") {
```
`data[i] == '/'`

ilya-kadochnikov commented on homework_01/task_02/src/utils.cpp:
```diff
@@ -1,7 +1,101 @@
+            continue;
+        }
+
+        if (data[i] == "-") {
```
`data[i] == '-'`

ilya-kadochnikov commented on homework_01/task_02/src/utils.cpp:
```diff
@@ -1,7 +1,101 @@
+            continue;
+        }
+
+        if (data[i] == "+") {
```
`data[i] == '+'`

ilya-kadochnikov commented on homework_01/task_02/src/utils.cpp:
```diff
@@ -1,7 +1,101 @@
+            continue;
+        }
+
+        v1.push_back(std::stoi(data[i]));
```
`v1.push_back(data[i] - '0');`

tatiana-kornienko14 committed:
Update utils.cpp

поправила названия переменных, добавила проверку на табуляцию

tatiana-kornienko14 committed:
Update utils.cpp

я так посмотрела, define у меня в принципе нигде не используется... упс
увидела, кстати, у себя заметную проблему в коде, скоро исправлю, заодно доисправляю названия переменных

Tatyana145 committed:
поменяла названияя переменных, скоро поправлю еще кое-что

Tatyana145 committed:
почти доделала, осталось немного

Tatyana145 committed:
вроде все поправила

Tatyana145 committed:
доделала; вот это уже должно работать

Tatyana145 committed:
обработала ошибку в main.cpp

Tatyana145 committed:
это не особо важно

Tatyana145 committed:
добавила условие со скобками, надо еще доработать

Tatyana145 committed:
вроде добила

Tatyana145 committed:
пара поправок

LostPointer commented on homework_01/task_01/src/utils.cpp:
```diff
@@ -3,5 +3,44 @@
+        is_bracket = 0;
+        if (tmp != "") {
+          answer.push_back(tmp);
+          tmp = "";
```
лучше оспользовать метод clear: ```tmp.clear()```
это даст немного больше понимания что ты очищаешь переменную, а не присваиваешь значение переменной

LostPointer commented on homework_01/task_02/src/utils.cpp:
```diff
@@ -1,7 +1,118 @@
+#include <iostream>
+#include <regex>
+#include <stack>
 #include "utils.hpp"
```
а вот если у тебя два файла с одинаковым мененм и расширениями .cpp и .hpp, в cpp инклуд hpp должен идти первым. получится примерно вот так:
```
#include "utils.hpp"

#include <regex>
#include <stack>
#include <iostream>
```

LostPointer commented on homework_01/task_02/src/utils.cpp:
```diff
@@ -1,7 +1,118 @@
+        }
+
+        if (!is_number(number)) {
+            throw ("ошибка!\n");
```
```throw std::runtime_error(number + " isn't numeber")

LostPointer commented on homework_01/task_02/src/utils.cpp:
```diff
@@ -1,7 +1,118 @@
+        }
+
+        if (multipl) {
+            terms[terms.size()-1] *= std::stoi(number);
```
пробелы пропущены: ```terms.size() - 1```

LostPointer reviewed (changes requested):
поправь в тестах для первого задания с { "a", "(a a)", "b", "(asd as)" } на { "a", "a a", "b", "asd as" }, так будут проходить тесты

Tatyana145 committed:
поправила

Tatyana145 committed:
поправила

ilya-kadochnikov commented on homework_01/task_02/src/utils.cpp:
```diff
@@ -1,7 +1,119 @@
+            number += data[i];
+            continue;
+        }
+
```
Скобочки у вас пока ещё корректно не обрабатываются. Наличие скобочек приводит к вылетам программы. Чтобы исправить вылеты, вставьте сюда код (но это ещё не всё):
```
if (number.empty())
  continue;
```

ilya-kadochnikov commented on homework_01/task_02/src/utils.cpp:
```diff
@@ -1,7 +1,119 @@
+        minus = 0;
+    } else {
+        terms.push_back(std::stoi(number));
+    }
```
Чтобы исправить вылеты из-за скобочек, окружите операторы if-else ещё одной проверкой:
```
if (!number.empty())
{
    if (multipl) {
        terms[terms.size() - 1] *= std::stoi(number);
        multipl = 0;
    } else if (division) {
        terms[terms.size() - 1] /= std::stoi(number);
        division = 0;
    } else if (minus) {
        terms.push_back(-std::stoi(number));
        minus = 0;
    } else {
        terms.push_back(std::stoi(number));
    }
}
```

ilya-kadochnikov commented on homework_01/task_02/src/utils.cpp:
```diff
@@ -1,7 +1,119 @@
+    
+    for (size_t i = 0; i < data.size(); ++i) {
+        if (is_brackets) {
+            if (data[i] == ')') {
```
Замечу просто, что если отключать флаг `is_brackets` сразу, как только встречается закрывающая скобка, то вложенные друг в друга скобки точно корректно оброватываться не будут. Впрочем, для первого задания и так хорошо

ilya-kadochnikov commented on homework_01/task_02/src/utils.cpp:
```diff
@@ -1,7 +1,119 @@
+        }
+
+        if (multipl) {
+            terms[terms.size() - 1] *= std::stoi(number);
```
`std::stoi(number)` возвращает `int`, а `term` хранит `double` (да и функция `Calculate`, вроде, должна `double` возвращать). Нестыковка)

Tatyana145 committed:
пока временно поменяла все на int, потом верну обратно; добавила empty

Tatyana145 committed:
СКОБОЧКИ РАБОТАЮТ, УРА :D а вообще осталось чуть-чуть

Tatyana145 committed:
теперь скобочки работают и внутри скобочек

Tatyana145 committed:
работает! :D
